### DISCUSSION TITLE
Rootless Podman Setup Throwing "attempt to write a readonly database"

### DISCUSSION THREAD

**Discussion by jollySleeper** [navidrome/navidrome#3309] [Q&A]
### I confirm that:

- [X] I have searched the existing [open AND closed issues](https://github.com/navidrome/navidrome/issues?q=is%3Aissue) to see if an issue already exists for the bug I've encountered
- [X] I'm using the latest version (your issue may have been fixed already)

### Version

v0.51.1

### Current Behavior

Hi, I have installed NaviDrome using Podman and am trying to run rootless with non-root user using the following command:
```bash
podman run \
 --restart unless-stopped \
 --label io.containers.autoupdate=registry \
 --user $(id -u):$(id -g) \
 --env-file "$HOME/navidrome/local.env" \
 -p 127.0.0.1:4533:4533 \
 -v "$HOME/navidrome/data":/data \
 -v ${HOME}/media/music:/music \
 --name "navidrome" \
 docker.io/deluan/navidrome;
```

But am getting the following error. 
```
time="2024-04-22T16:47:26Z" level=error msg="Error disabling foreign_keys" error="attempt to write a readonly database"
time="2024-04-22T16:47:26Z" level=fatal msg="Database could not be opened!" error="attempt to write a readonly database"
```

I have seen same issues reported by other users but did not find any solution. 
Does Navidrome doesn't support running with Non-Root User??

PS: 
- Output of `$(id -u)` gives `1001` as my user-id is `1001` and not `1000`.
- And the `data` folder is owned by user.

### Expected Behavior

Should Work

### Steps To Reproduce

As Written Above

### Environment

```markdown
- OS: Ubuntu 24.04
- Browser: 
- Client: ?
```

### How Navidrome is installed?

Docker

### Code of Conduct

- [X] I agree to follow Navidrome's Code of Conduct

**Comment by deluan** [comment]
> Does Navidrome doesn't support running with Non-Root User??

Not completely sure about Podman, but for Docker: Yes it does. But you have to make sure that the `/data` volume is writable by the userid you choose. 

If the path you choose to mount as `/data` does not exist when you first start the container, Docker will create it with its own permissions (usually `root`), and then mount it in the container. That causes Navidrome (running under the userid you choose) to not have write permissions on the `/data` folder (owned by `root`), leading to the errors you are seeing.

You can try doing this:
- Stop the container
- `chown $(id -u):$(id -g) -R $HOME/navidrome/data`
- Start the container

If it is not working with Podman after this, I can't really help you as I don't use nor have an installation of Podman, and you better off reaching out to the community in Reddit, Discord or our GH [Discussions](https://github.com/navidrome/navidrome/discussions/categories/q-a)

**Comment by jollySleeper** [comment]
Yes I tried Multiple times before drafting this Issue. Even changed the ownership and gave 777 to db but just to get the same error again :(

Will try out more.

**Comment by deluan** [comment]
Try removing the `$HOME/navidrome/data` folder, and creating it with the correct permissions in the first place.

After doing this, I can get it running with my userid:

```
$ mkdir $HOME/data
$ docker run  --user $(id -u):$(id -g)  -p 127.0.0.1:4533:4533  -v "$HOME/data":/data  -v ${HOME}/music:/music  --name "navidrome" docker.io/deluan/navidrome
 _   _             _     _
| \ | |           (_)   | |
|  \| | __ ___   ___  __| |_ __ ___  _ __ ___   ___
| . ` |/ _` \ \ / / |/ _` | '__/ _ \| '_ ` _ \ / _ \
| |\  | (_| |\ V /| | (_| | | | (_) | | | | | |  __/
\_| \_/\__,_| \_/ |_|\__,_|_|  \___/|_| |_| |_|\___|
                          Version: 0.51.1 (6d253225)

time="2024-04-22T21:56:32Z" level=info msg="Creating DB Schema"
time="2024-04-22T21:56:33Z" level=info msg="Starting signaler"
time="2024-04-22T21:56:33Z" level=error msg="Agent not available. Check configuration" name=lastfm
time="2024-04-22T21:56:33Z" level=error msg="Agent not available. Check configuration" name=spotify
time="2024-04-22T21:56:33Z" level=info msg="Configuring Media Folder" name="Music Library" path=/music
time="2024-04-22T21:56:33Z" level=info msg="Creating Image cache" maxSize="100 MB" path=/data/cache/images
time="2024-04-22T21:56:33Z" level=info msg="Starting scheduler"
time="2024-04-22T21:56:33Z" level=info msg="Scheduling periodic scan" schedule="@every 1m"
time="2024-04-22T21:56:33Z" level=info msg="Running initial setup"
time="2024-04-22T21:56:33Z" level=info msg="Creating new JWT secret, used for encrypting UI sessions"
time="2024-04-22T21:56:33Z" level=info msg="Setting Session Timeout" value=24h
time="2024-04-22T21:56:33Z" level=info msg="Login rate limit set" requestLimit=5 windowLength=2
time="2024-04-22T21:56:33Z" level=info msg="Finished initializing cache" cache=Image elapsedTime=263.3ms maxSize=100MB
time="2024-04-22T21:56:34Z" level=info msg="Found ffmpeg" path=/usr/bin/ffmpeg
time="2024-04-22T21:56:34Z" level=info msg="Spotify integration is not enabled: missing ID/Secret"
time="2024-04-22T21:56:34Z" level=info msg="Mounting Native API routes" path=/api
time="2024-04-22T21:56:34Z" level=error msg="Agent not available. Check configuration" name=lastfm
time="2024-04-22T21:56:34Z" level=error msg="Agent not available. Check configuration" name=spotify
time="2024-04-22T21:56:34Z" level=info msg="Creating Transcoding cache" maxSize="100 MB" path=/data/cache/transcoding
time="2024-04-22T21:56:34Z" level=info msg="Finished initializing cache" cache=Transcoding elapsedTime="337.1µs" maxSize=100MB
time="2024-04-22T21:56:34Z" level=info msg="Mounting Subsonic API routes" path=/rest
time="2024-04-22T21:56:34Z" level=error msg="Agent not available. Check configuration" name=lastfm
time="2024-04-22T21:56:34Z" level=error msg="Agent not available. Check configuration" name=spotify
time="2024-04-22T21:56:34Z" level=info msg="Mounting Public Endpoints routes" path=/share
time="2024-04-22T21:56:34Z" level=info msg="Mounting LastFM Auth routes" path=/api/lastfm
time="2024-04-22T21:56:34Z" level=info msg="Mounting ListenBrainz Auth routes" path=/api/listenbrainz
time="2024-04-22T21:56:34Z" level=info msg="Mounting Background images routes" path=/backgrounds
time="2024-04-22T21:56:34Z" level=info msg="Mounting WebUI routes" path=/app
time="2024-04-22T21:56:34Z" level=info msg="----> Navidrome server is ready!" address="0.0.0.0:4533" startupTime=6.78s tlsEnabled=false
```

**Comment by andrewdunndev** [comment]
Have selinux on your system? your volume mounts should have `:z` or `:Z` appended to them if so.

**Comment by jcleng** [comment]
when id not root user cant runing 
```
id -u
# 1000

docker run -itd \
   --name navidrome \
   --restart=unless-stopped \
   --user $(id -u):$(id -g) \
....

FATAL: Error creating cache path: path /data/cache mkdir /data/cache: permission denied
```

**Comment by jacktom12** [comment]
chown 1000:1000 -R also good work

**Comment by emuhr** [comment]
I also received this error. Creating the data directory with the correct ownership and permissions did not work.
What works for me is to change ownership inside a [user namespace](https://docs.podman.io/en/latest/markdown/podman-unshare.1.html):
```podman unshare chown -R 1000:1000 ./data```

**Comment by kgarner7** [comment]
Note that with rootless podman you _most likely_ have user namespace remapping (also available in Docker): https://github.com/containers/podman/blob/main/docs/tutorials/rootless_tutorial.md

In which case, the id in the container would be remapped to the uid/gid in `/etc/subuid`/`/etc/subgid`. As an example, if you have the following
```bash
# cat /etc/subuid
johndoe:100000:65536
test:165536:65536
```
Then id `1000` in the container would be remapped to id `101000` on the host.

**Comment by deluan** [comment]
This issue documented quite a few ways to fix/resolve this issue. It is not really a Navidrome issue, so what do you folks think of converting this issue to a Q&A discussions?

**Comment by kgarner7** [comment]
> This issue documented quite a few ways to fix/resolve this issue. It is not really a Navidrome issue, so what do you folks think of converting this issue to a Q&A discussions?

Honestly, might be worthwhile
1. Making a discussion
2. Documenting rootless behavior (Docker/Podman) in the docs. (or at least say "if you are using rootless read these other docs")

**Comment by rdapgar** [comment]
I would like to share the solution I am using with podman rootless. Create a docker-compose.yml file like:
```
version: "3"
services:
  navidrome:
    image: deluan/navidrome:latest
    userns_mode: keep-id
    ports:
      - "4533:4533"
    restart: unless-stopped
    environment:
      # Change 1000 in PODMAN_USERNS to your current UID(id -u) and GID(id -g)
      PODMAN_USERNS: "keep-id:uid=1000:999,gid=1000:999"
      # Optional: put your config options customization here. Examples:
      ND_SCANSCHEDULE: 1h
      ND_LOGLEVEL: info  
      ND_SESSIONTIMEOUT: 24h
      ND_BASEURL: ""
    volumes:
      - "/path/to/data:/data:U"
      - "/path/to/your/music/folder:/music:ro"
```
Then, simply a "podman-compose up -d" command should work.

**Comment by Tootam** [comment]
if someone is still struggling with rootless docker - setting user to 0:0 helped (not sure how valid it is but it finally works):
```
services:
    navidrome:
      image: deluan/navidrome:latest
      user: 0:0
      ports:
        - 4533:4533
      restart: unless-stopped
      volumes:
        - "/example/path/data:/data"
        - "/example/path/music:/music:ro"
```
setting user to what was returned by
id -u 
id -g 
caused multiple issues with write permissions/initial db setup

**Comment by pierrecnalb** [comment]
In case anyone is interested, I created [here](https://github.com/pierrecnalb/quadlets/tree/main/navidrome) some hardened navidrome quadlets with the most secure options for server deployment.
As described in the repository, to achieve the maximum security, the containers are using the following options:

- a pod is used to host navidrome containers
- the pod is properly isolated within its own user namespace (`UserNS=auto`)
- containers are running with a dedicated user (`User=...`)
- the least amount of privileges are used (`NoNewPrivileges=true`, `ReadOnly=true` etc)
- all capabilities removed (`DropCapability=all`)
